paramiko/pkey.py

```diff
@@ -39,6 +39,107 @@
 from paramiko.message import Message
 
 
+_legacy_type_registry = {}
+_openssh_type_registry = []
+
+
+def register_pkey_type(cls):
+    """Decorator for PKey subclasses to register their types for parsing."""
+    if cls.LEGACY_TYPE:
+        _legacy_type_registry[cls.LEGACY_TYPE] = cls
+    if cls.OPENSSH_TYPE_PREFIX:
+        _openssh_type_registry.append((cls.OPENSSH_TYPE_PREFIX, cls))
+    return cls
+
+
+# TODO: look at what documentation needs to be written
+# TODO: look for code in paramiko-ng internally that could use this
+# TODO: check code style with Black?
```

Comment:
I'm not a fan of Black, and forked from upstream paramiko just before it was run across the codebase. I prefer just flake8/pycodestyle.

The main place inside paramiko that could benefit from these two new functions is `SSHClient._auth()`
https://github.com/ploxiln/paramiko-ng/blob/2.7.4/paramiko/client.py#L593-L598

Reply:
Right, I'll skip the Black part then. :)

Not trying to push it or anything, just FWIW: I started out not liking the Black style and auto-formatting in general. But after trialing it for the past year on a work project, it completely changed my mind. I feel like it frees up cognitive resources that were previously occupied by code formatting (in those projects I configure my IDE to auto-format on every save). It immediately shuts down any bikeshedding over space placement and whatnot. And new people on a project don't need to spend time on learning the project's particular code style, especially relevant for open source projects.

Adopting Black would also simplify porting patches between `paramiko` and `paramiko-ng`. But that doesn't seem to be a big issue in practice; resolving rebase conflicts probably took me 10 minutes on this original patch.

Reply:
I don't use an IDE. I make sure that my editor does not modify whitespace (or anything) on any lines which I did not intentionally edit.

The auto-formatter can't fix bad variable and function names, or bad code organization. And it gets some things obviously wrong. Trying to make the auto-formatter correct in 100% of cases is a much more challenging problem than just writing the code of interest with reasonable formatting manually. The churn to the codebase makes "git blame" much more annoying. I ported a bunch of PRs from upstream paramiko already (and most were pre-Black).

"go fmt" isn't so bad because it's a bit more like flake8 - it enforces indentation and other kinds of spacing, but lets you wrap lines in a few different ways. (But I still find it's choice of whitespace around nested arithmetic operators sometimes quite irritating.)

Reply:
> I make sure that my editor does not modify whitespace

That's explicitly not a use case of Black. Black is all-or-nothing: either everybody working on a project uses it, or nobody can reasonably use it.

> Trying to make the auto-formatter correct in 100% of cases

I assume by "correct" you mean "looks pleasing"? I agree sometimes there are cases where you have to help it along a bit, by extracting sub-expressions into variables. But that also tends to increase readability in general.

I think Black is mostly being held back by the obsolete line length limit of 88 chars (or worse, 79 in upstream paramiko). My preference in personal projects is 120 (with or without Black). `--line-length=120` gets rid of a lot of unnecessary line splitting with Black, at which point while sometimes worse than hand-tuned formatting, in some cases I find it's better than what I see humans produce, and ends up roughly on par. With the added benefit of absolute consistency, that's still a win.

> The churn to the codebase makes "git blame" much more annoying.

If everybody uses Black's code style then I find it actually reduces unnecessary churn (except for the initial conversion at first).